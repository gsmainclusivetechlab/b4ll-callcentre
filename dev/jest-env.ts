import env from 'dotenv';

env.config({ path: __dirname + '/../jest.env' });
