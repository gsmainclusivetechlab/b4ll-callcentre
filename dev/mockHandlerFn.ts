/* istanbul ignore file */

import { safeHandle, ParsedRequest } from '../src/services/safeHandle';

type DeepPartial<T> = {
    [P in keyof T]?: DeepPartial<T[P]>;
};
type Handler = Parameters<typeof safeHandle>[0];
type TestInput = Omit<ParsedRequest, 'event' | 'auth'> & {
    event?: DeepPartial<ParsedRequest['event']>;
    auth?: ParsedRequest['auth'];
};
type HandlerResult = ReturnType<Handler>;

/**
 * Undoes the effect of the `safeHandle` function, and mocks the event parameter
 * @param handler
 */
export function mockHandlerFn(
    handler: unknown
): (input: TestInput) => HandlerResult {
    if (
        typeof handler === 'function' &&
        handler.name === 'safelyHandledFunction' &&
        'orig' in handler
    ) {
        const orig = (handler as Record<string, unknown>)['orig'];
        if (typeof orig === 'function') {
            return ({ event, ...params }) =>
                orig({ ...params, event: event || {} });
        }
    }
    throw new Error("Can't mock handlers which are not safely handled");
}
