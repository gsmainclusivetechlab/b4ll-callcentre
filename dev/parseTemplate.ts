import { readFileSync } from 'fs';
import { yamlParse } from 'yaml-cfn';
import * as t from 'io-ts';
import { either } from 'fp-ts';
import { pipe } from 'fp-ts/lib/function';

function substituteVars(s: string) {
    return s.replace(/\$\{([^}]+)\}/g, (orig, v) => {
        switch (v) {
            case 'Namespace':
                return process.env.STAGE || 'jest';
        }
        return orig;
    });
}

const Subbable = new t.Type<string, string, unknown>(
    'string',
    (input: unknown): input is string =>
        typeof input === 'string' ||
        (typeof input === 'object' && input !== null && 'Fn::Sub' in input),
    (input, context) => {
        if (typeof input === 'string') return t.success(input);
        if (typeof input === 'object' && input !== null && 'Fn::Sub' in input) {
            const i = input as { 'Fn::Sub': unknown };
            if (typeof i['Fn::Sub'] === 'string')
                return t.success(substituteVars(i['Fn::Sub']));
        }
        return t.failure(input, context);
    },
    t.identity
);

const ServerlessFunction = t.type({
    Type: t.literal('AWS::Serverless::Function'),
    Properties: t.type({
        Runtime: t.union([t.undefined, t.string]),
        Handler: t.string,
        CodeUri: t.string,
    }),
});
export type ServerlessFunctionTemplate = t.TypeOf<typeof ServerlessFunction>;

const DynamoDBTable = t.type({
    Type: t.literal('AWS::DynamoDB::Table'),
    Properties: t.type({
        TableName: Subbable,
        AttributeDefinitions: t.array(
            t.type({
                AttributeName: t.string,
                AttributeType: t.keyof({ S: null, N: null }),
            })
        ),
        KeySchema: t.array(
            t.type({
                AttributeName: t.string,
                KeyType: t.keyof({ HASH: null, RANGE: null }),
            })
        ),
        ProvisionedThroughput: t.type({
            ReadCapacityUnits: t.number,
            WriteCapacityUnits: t.number,
        }),
    }),
});
export type DynamoDBTableTemplate = t.TypeOf<typeof DynamoDBTable>;

const CloudFormationSchema = t.intersection([
    // optional properties
    t.partial({
        Globals: t.partial({
            Function: t.partial({
                Runtime: t.string,
            }),
        }),
    }),
    t.type({
        Resources: t.record(
            t.string,
            t.union([
                ServerlessFunction,
                DynamoDBTable,
                t.type({
                    Type: t.string,
                }),
            ])
        ),
    }),
]);
export type CloudFormationTemplate = t.TypeOf<typeof CloudFormationSchema>;

export function getTemplate(templatePath: string): CloudFormationTemplate {
    const parseResult = pipe(
        either.tryCatch(
            () => yamlParse(readFileSync(templatePath).toString()),
            // ignore validation errors for now (TODO: nicer errors)
            (e) => {
                console.error(e);
                return []
            }
        ),
        either.chain(CloudFormationSchema.decode)
    );

    if (either.isLeft(parseResult)) {
        console.log(parseResult.left)
        throw new Error('Unable to parse yaml');
    }
    return parseResult.right;
}
