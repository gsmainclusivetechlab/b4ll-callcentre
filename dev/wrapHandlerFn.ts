import {
    APIGatewayProxyHandlerV2,
    APIGatewayProxyEventV2,
    APIGatewayProxyResultV2,
    Context,
} from 'aws-lambda';

export function wrapHandlerFn(
    handler: APIGatewayProxyHandlerV2
): (
    props: Partial<APIGatewayProxyEventV2>
) => void | Promise<APIGatewayProxyResultV2> {
    return (props) =>
        handler(
            props as APIGatewayProxyEventV2,
            {} as Context,
            () => ({})
        );
}
