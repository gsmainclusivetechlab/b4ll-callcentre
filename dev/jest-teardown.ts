import './jest-env';
import { down } from 'docker-compose';

module.exports = async () => {
    process.stdout.write('Stopping DynamoDB... ');
    await down({ cwd: __dirname });
    console.log('[Done]');
};
