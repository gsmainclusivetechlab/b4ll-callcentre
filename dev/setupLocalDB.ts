import { upOne } from 'docker-compose';
import AWS from 'aws-sdk';
import { getTemplate, DynamoDBTableTemplate } from './parseTemplate';

export async function setupLocalDDB(): Promise<void> {
    const ddbPromise = upOne('dynamodb', { cwd: __dirname });
    const cfn = getTemplate(__dirname + '/../template.yaml');
    const tables = Object.keys(cfn.Resources)
        .filter((k) => cfn.Resources[k].Type === 'AWS::DynamoDB::Table')
        .map((name) => {
            const r = cfn.Resources[name] as DynamoDBTableTemplate;
            return {
                ...r,
                Properties: {
                    ...r.Properties,
                    TableName: r.Properties.TableName || name,
                },
            };
        });
    await ddbPromise.then(async () => {
        process.stdout.write('\nSetting up test tables... ');
        const db = new AWS.DynamoDB({
            endpoint: `http://localhost:${process.env.DDB_PORT || 8000}`,
            sslEnabled: false,
            region: 'local-env',
        });
        const currentTables =
            (await db.listTables().promise()).TableNames || [];
        await Promise.all(
            tables.map(async (t, i) => {
                const TableName = t.Properties.TableName || `Table${i}`;
                if (currentTables.indexOf(TableName) >= 0) {
                    // table was already created - let's remove it to start afresh
                    await db
                        .deleteTable({
                            TableName,
                        })
                        .promise();
                }
                return db.createTable({ ...t.Properties, TableName }).promise();
            })
        );
        console.log('[Done]');
    });
}

if (require.main === module) {
    // if called directly from the command line, execute now
    setupLocalDDB();
}
