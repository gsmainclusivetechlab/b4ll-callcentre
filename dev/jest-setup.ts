import "./jest-env";
import { setupLocalDDB } from './setupLocalDB';

module.exports = async () => {
    await setupLocalDDB();
};
