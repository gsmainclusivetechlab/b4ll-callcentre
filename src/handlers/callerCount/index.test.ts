import * as orig from '.';
import { mockHandlerFn } from '../../../dev/mockHandlerFn';

const get = mockHandlerFn(orig.get);

describe('Caller count', () => {
    it('should return well-formed XML', async () => {
        const result = await get({
            language: 'en-GB',
            user: { id: '+77-caller-test' },
        });
        expect(result.toString()).toMatchSnapshot();
    });

    it('should increment caller count', async () => {
        const result = await get({
            language: 'en-GB',
            user: { id: '+77-caller-test', count: 1 },
        });
        expect(result.toString()).toMatchSnapshot();
    });
});
