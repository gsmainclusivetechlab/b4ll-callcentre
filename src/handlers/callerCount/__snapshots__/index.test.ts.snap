// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caller count should increment caller count 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Emma-Neural">
        Thanks for calling. This was the 2nd time you've called.
    </Say>
</Response>
`;

exports[`Caller count should return well-formed XML 1`] = `
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say language="en-GB" voice="Polly.Emma-Neural">
        Thanks for calling. This was the 1st time you've called.
    </Say>
</Response>
`;
